umapi_client/functional.py

```diff
@@ -196,14 +210,18 @@ def update(self, email=None, username=None, first_name=None, last_name=None, cou
         :return: the User, so you can do User(...).update(...).add_to_groups(...)
         """
         if email:
-            self._validate(email=email)
-        if username:
-            self._validate(username=username)
+            self._validate_email(email=email)
+        # if username contains @ (federated only), validate as email address
```

Comment:
OK, I see, you expect to decide which validator to call, rather than updating the validator to be permissive.  But you only did this on update, and you forgot init.  I think just making the validator more permissive is a better idea, then you don't need special-case code anywhere, and you don't have to check to see whether it's Federated ID or not.

Reply:
I didn't forget init.  The rules don't change on init for validating email/username.  When creating a user, the email address and username must match if the username is an email address.  Only when updating a user are the two fields allowed to differ.

That's why I chose to maintain separate validation functions instead of making the username validator more permissive.  There are still circumstances where we need to validate a username as a non-email username.

The Federated ID check is there because I don't believe Enterprise IDs are allowed to have different email addresses and usernames.  Do you know if that's allowed?  I thought I tested it and confirmed that it isn't.  I'll test it again.

Reply:
No, there has *never* been a requirement that usernames match email, even for federated where the username is in email form.  For a long time there was a requirement that the domain of an email-formatted username had to match the domain of the user's email, but that was a bogus requirement and we finally got it changed about 4 months ago.  So you really need to relax this validation requirement.